### PR TITLE
Lazily evaluate Ingress address

### PR DESCRIPTION
**Pull request by mandarjog**
Ingress takes up to 2 minutes to get an address.
We could be doing other work during that time.

This PR updates code to wait for ingress address only when the code is about to use that address.

### CODE REVIEW

**Comment by ldemailly** [comment]
in theory it's great to do as much stuff as possible in parallel but getting the ingress really up is gating every e2e tests anyway so I'm not sure of the gain here - we just discussed in the other PR (https://github.com/istio/istio/pull/4050#issuecomment-371666849) that we need to do more ingress checks before spending time starting other tests doomed to fail

**Comment by mandarjog** [comment]
Ingress gates the test, but it does not getting setting up the applications.
There are tests that do `kubeexec` and do not do use ingress. If those are running first, you don't need to wait for ingress adress to appear before starting tests.  I expect it to shave off  a minute or two from these tests.

**Comment by ldemailly** [comment]
that or our ingress could come up in 10s like it should :-) but ok, thanks for this! (if it works and doesn't have weird side effects)

**Comment by ldemailly** [comment]
want to test your theory by swapping the order of some tests ?

**Comment by mandarjog** [comment]
Old test `173 sec` from start of process to starting running real test
```
Mixer test starts
2018-03-07T21:33:16.890210Z	info	Logging initialized
.
2018-03-07T21:34:45.042793Z	info	Istio ingress: 35.227.55.149
.
2018-03-07T21:36:03.786502Z	info	Get all pods running!
2018-03-07T21:36:05.650886Z	info	Get from http://35.227.55.149/productpage: 200 OK (200)
2018-03-07T21:36:05.651228Z	info	Got 200 response from product page!
```

New test `120 sec`
```
I0309 00:20:02.799] 2018-03-09T00:20:02.799287Z	info	Logging initialized
.
.
I0309 00:22:00.560] 2018-03-09T00:22:00.560314Z	info	Istio ingress: 35.224.39.44
I0309 00:22:02.218] 2018-03-09T00:22:02.217755Z	info	Get from http://35.224.39.44/productpage: 200 OK (200)
I0309 00:22:02.218] 2018-03-09T00:22:02.217951Z	info	Got 200 response from product page!
```

**Comment by mandarjog** [comment]
@costinm @ldemailly all jobs are succeeding. /cc @ayj

File: install/updateVersion.sh [hunk 1]
```diff
@@ -173,6 +173,7 @@ function merge_files() {
   execute_sed "s/# authPolicy: MUTUAL_TLS/authPolicy: MUTUAL_TLS/" $ISTIO_AUTH
   execute_sed "s/# controlPlaneAuthPolicy: MUTUAL_TLS/controlPlaneAuthPolicy: MUTUAL_TLS/" $ISTIO_AUTH
   execute_sed "s/NONE #--controlPlaneAuthPolicy/MUTUAL_TLS/" $ISTIO_AUTH
+  execute_sed "s/8080 #--controlPlaneAuthPolicy/15005/" $ISTIO_AUTH
```
Comment (ldemailly) [inline install/updateVersion.sh]: these changes look like from the auth fix PR - is this PR not rebased or some such ?

**Comment by ldemailly** [comment]
please rebase, this seemingly includes #4066 
/hold

**Comment by mandarjog** [comment]
@ldemailly I had cherry picked those changes while the other PR was still in flight.
When you merge into master those changes go away.
Any way I have rebased.

File: tests/e2e/framework/kubernetes.go [hunk 2]
```diff
@@ -178,6 +171,39 @@ func (k *KubeInfo) Setup() error {
 	return nil
 }
 
+// IngressOrFail lazily initialize ingress and fail test if not found.
+func (k *KubeInfo) IngressOrFail(t *testing.T) string {
+	gw, err := k.Ingress()
+	if err != nil {
+		t.Fatalf("Unable to get ingress: %v", err)
```
Comment (ldemailly) [inline tests/e2e/framework/kubernetes.go]: does it dump the ingress logs and/or pod status when that happens?

**Comment by ldemailly** [comment]
thx
/lgtm
what fails on circle now ?

**Comment by ayj** [comment]
```
0309 19:02:50.404] --- FAIL: TestVerifyCert (0.00s)
I0309 19:02:50.405] 	verify_cert_test.go:367: Succeeded: Unexpected error: failed to verify certificate: x509: certificate has expired or is not yet valid
I0309 19:02:50.405] 	verify_cert_test.go:369: extKeyUsage error: Unexpected error: failed to verify certificate: x509: certificate has expired or is not yet valid VS (expected) unexpected value for 'ExtKeyUsage' field
I0309 19:02:50.405] 	verify_cert_test.go:369: KeyUsage Error: Unexpected error: failed to verify certificate: x509: certificate has expired or is not yet valid VS (expected) unexpected value for 'KeyUsage' field
I0309 19:02:50.405] 	verify_cert_test.go:369: Org error: Unexpected error: failed to verify certificate: x509: certificate has expired or is not yet valid VS (expected) unexpected value for 'Organization' field
I0309 19:02:50.406] 	verify_cert_test.go:369: TTL error: Unexpected error: failed to verify certificate: x509: certificate has expired or is not yet valid VS (expected) unexpected value for 'NotAfter' - 'NotBefore'
I0309 19:02:50.406] 	verify_cert_test.go:369: IsCA error: Unexpected error: failed to verify certificate: x509: certificate has expired or is not yet valid VS (expected) unexpected value for 'IsCA' field
I0309 19:02:50.407] 	verify_cert_test.go:369: Failed to verify key: Unexpected error: failed to verify certificate: x509: certificate has expired or is not yet valid VS (expected) invalid PEM-encoded key
I0309 19:02:50.407] 	verify_cert_test.go:369: Failed to match key/cert: Unexpected error: failed to verify certificate: x509: certificate has expired or is not yet valid VS (expected) the generated private key and cert doesn't match
I0309 19:02:50.407] 	verify_cert_test.go:369: Wrong SAN: Unexpected error: failed to verify certificate: x509: certificate has expired or is not yet valid VS (expected) the certificate doesn't have the expected SAN for: spiffe
I0309 19:02:50.408] 	verify_cert_test.go:369: Timestamp error: Unexpected error: failed to verify certificate: x509: certificate has expired or is not yet valid VS (expected) unexpected value for 'NotBefore' field
```

@wattli is working on a fix.

**Pull request merged**